tests/test_sparse_matrix_device_operator.cu

```diff
@@ -0,0 +1,145 @@
+  matrix_operator.apply(*domain_dev, *range_dev);
+  dealii::Vector<double> domain_dealii(domain_vector.begin(),
+                                       domain_vector.end());
+  dealii::Vector<double> range_dealii(n_rows);
+  sparse_matrix.vmult(range_dealii, domain_dealii);
+  std::vector<double> range_vector(n_rows);
+  mfmg::cuda_mem_copy_to_host(range_dev->val_dev, range_vector);
+  for (unsigned int i = 0; i < n_rows; ++i)
+    BOOST_CHECK_EQUAL(range_vector[i], range_dealii[i]);
+
+  // Check transpose
+  dealii::SparsityPattern transpose_sparsity_pattern;
+  std::vector<std::vector<unsigned int>> transpose_column_indices(n_cols);
+  for (unsigned int i = 0; i < n_cols; ++i)
+    for (unsigned int j = 0; j < n_rows; ++j)
+      if (sparsity_pattern_indices.count(std::make_pair(j, i)) > 0)
+        transpose_column_indices[i].emplace_back(j);
+  transpose_sparsity_pattern.copy_from(n_cols, n_rows,
+                                       transpose_column_indices.begin(),
+                                       transpose_column_indices.end());
+  dealii::SparseMatrix<double> transpose_sparse_matrix(
+      transpose_sparsity_pattern);
+  for (unsigned int i = 0; i < n_cols; ++i)
+    for (unsigned int j = 0; j < n_rows; ++j)
+      if (sparsity_pattern_indices.count(std::make_pair(j, i)) > 0)
+        transpose_sparse_matrix.set(i, j, static_cast<double>(i + j));
+  auto transpose_matrix_operator = matrix_operator.transpose();
+  BOOST_CHECK_EQUAL(transpose_matrix_operator->m(),
+                    transpose_sparse_matrix.m());
+  BOOST_CHECK_EQUAL(transpose_matrix_operator->n(),
+                    transpose_sparse_matrix.n());
+  // We don't have access to the underlying matrix directly so we apply the new
+  // operator and check the results
+  auto transpose_domain_dev = transpose_matrix_operator->build_domain_vector();
+  auto transpose_range_dev = transpose_matrix_operator->build_range_vector();
+  std::vector<double> transpose_domain_vector(transpose_matrix_operator->n(),
+                                              1.);
+  mfmg::cuda_mem_copy_to_dev(transpose_domain_vector,
+                             transpose_domain_dev->get_values());
+  transpose_matrix_operator->apply(*transpose_domain_dev, *transpose_range_dev);
+  dealii::Vector<double> transpose_domain_dealii(
+      transpose_domain_vector.begin(), transpose_domain_vector.end());
+  dealii::Vector<double> transpose_range_dealii(transpose_matrix_operator->m());
+  transpose_sparse_matrix.vmult(transpose_range_dealii,
+                                transpose_domain_dealii);
+  std::vector<double> transpose_range_vector(transpose_matrix_operator->m());
+  mfmg::cuda_mem_copy_to_host(transpose_range_dev->val_dev,
+                              transpose_range_vector);
+  for (unsigned int i = 0; i < n_rows; ++i)
+    BOOST_CHECK_EQUAL(transpose_range_vector[i], transpose_range_dealii[i]);
+
+  // Check multiply
+  sparse_matrix.vmult(range_dealii, transpose_range_dealii);
```

Comment:
Similar to above, I would like to have explicit calls to random here to make sure.